pkg/kubelet/kubelet.go

```diff
@@ -905,14 +905,17 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 			}
 			if api.IsPullAlways(container.ImagePullPolicy) ||
 				(api.IsPullIfNotPresent(container.ImagePullPolicy) && (!present || latest)) {
+				kl.pullLock.RLock()
 				if err := kl.dockerPuller.Pull(container.Image); err != nil {
 					if ref != nil {
 
 						record.Eventf(ref, "failed", "failed", "Failed to pull image %q", container.Image)
 					}
 					glog.Errorf("Failed to pull image %q: %v; skipping pod %q container %q.", container.Image, err, podFullName, container.Name)
+					kl.pullLock.RUnlock()
```

Comment:
Don't you need to hold the lock until the container has been started?  Otherwise teh GC could delete it again between your unlock and your run?